--- bftpd/commands.c	2022-01-29 01:55:18.000000000 +0800
+++ bftpd-nsfuzz-patch/commands.c	2026-04-12 14:49:20.265429518 +0800
@@ -61,19 +61,19 @@
 #undef WANT_GZIP
 #endif
 
-int state = STATE_CONNECTED;
+int _NSFUZZ_STATE(state) = STATE_CONNECTED;
 char user[USERLEN + 1];
 struct sockaddr_in sa;
-char pasv = 0;
+char _NSFUZZ_STATE(pasv) = 0;
 int sock;
 int pasvsock = -1;
 char *philename = NULL;
 unsigned long offset = 0;
-short int xfertype = TYPE_BINARY;
+short int _NSFUZZ_STATE(xfertype) = TYPE_BINARY;
 int ratio_send = 1, ratio_recv = 1;
 /* long unsigned bytes_sent = 0, bytes_recvd = 0; */
 double bytes_sent = 0.0, bytes_recvd = 0.0;
-int epsvall = 0;
+int _NSFUZZ_STATE(epsvall) = 0;
 int xfer_bufsize;
 
 
@@ -293,7 +293,7 @@
   sa.sin_addr.s_addr = addr;
   sa.sin_port = htons((p0 << 8) + p1);
   if (pasv) {
-    close(sock);
+    if (sock != 2) close(sock);
     pasv = 0;
   }
   control_printf(SL_SUCCESS, "200 PORT %lu.%lu.%lu.%lu:%lu OK",
@@ -331,7 +331,7 @@
     }
     sa.sin_port = htons(port);
     if (pasv) {
-        close(sock);
+        if (sock != 2) close(sock);
         pasv = 0;
     }
     control_printf(SL_FAILURE, "200 EPRT %s:%i OK", addr, port);
@@ -506,7 +506,7 @@
     if ( (result) &&  (strstr(str, "ABOR")) ) {
         control_printf(SL_SUCCESS, "426 Transfer aborted.");
     	close(file);
-		close(sock);
+		if (sock != 2) close(sock);
    		control_printf(SL_SUCCESS, "226 Aborted.");
 		bftpd_log("Client aborted file transmission.\n");
         alarm(control_timeout);
@@ -642,7 +642,7 @@
                      "553 Error: Remote file is write protected.");
 
               free(mapped);
-              close(sock);
+              if (sock != 2) close(sock);
               return;
            }
         }
@@ -680,7 +680,7 @@
            if (! my_zip_file)
            {
               control_printf(SL_FAILURE, "553 Error: An error occurred creating compressed file.");
-              close(sock);
+              if (sock != 2) close(sock);
               close(fd);
               return;
            }
@@ -724,7 +724,7 @@
        control_printf(SL_FAILURE, "553 Error: An unknown error occurred on the server.");
        if (fd >= 0)
           close(fd);
-       close(sock);
+       if (sock != 2) close(sock);
        if (mapped)
           free(mapped);
        return;
@@ -745,7 +745,7 @@
         tv.tv_sec = data_timeout;
         tv.tv_usec = 0;
         if (!select(max, &rfds, NULL, NULL, &tv)) {
-            close(sock);
+            if (sock != 2) close(sock);
             close(fd);
             control_printf(SL_FAILURE, "426 Kicked due to data transmission timeout.");
             bftpd_log("Kicked due to data transmission timeout.\n");
@@ -828,7 +828,7 @@
         if (fd >= 0)
           close(fd);
 
-	close(sock);
+	if (sock != 2) close(sock);
         alarm(control_timeout);
         offset = 0;
 	control_printf(SL_SUCCESS, "226 File transmission successful.");
@@ -1273,7 +1273,7 @@
                         {
                             control_printf(SL_FAILURE, "553 An unknown error occurred.");
                             bftpd_log("Memory error while trying to send file.", 0);
-                            close(sock);
+                            if (sock != 2) close(sock);
                             close(phile);
                             return;
                         }
@@ -1285,7 +1285,7 @@
                         else
                             my_buffer_size = xfer_bufsize;
 
-                        i = read(phile, buffer, my_buffer_size);
+                        i = 0;//read(phile, buffer, my_buffer_size);
 			while (i > 0) {
 				if (test_abort(1, phile, sock)) {
 					free(buffer);
@@ -1302,7 +1302,7 @@
                                 {
                                    free(buffer);
                                    close(phile);
-                                   close(sock);
+                                   if (sock != 2) close(sock);
                                    alarm(control_timeout);
                                    control_printf(SL_SUCCESS, "426 Transfer aborted.");
                                    control_printf(SL_SUCCESS, "226 Aborted.");
@@ -1325,13 +1325,13 @@
                                    wait_time.tv_usec = xfer_delay;
                                    select( 0, NULL, NULL, NULL, &wait_time);
                                 }
-                                i = read(phile, buffer, my_buffer_size);
+                                i = 0;//read(phile, buffer, my_buffer_size);
 			}       // end of while
             free(buffer);
             }
 
 	close(phile);
-	close(sock);
+	if (sock != 2) close(sock);
         offset = 0;
         alarm(control_timeout);
 	control_printf(SL_SUCCESS, "226 File transmission successful.");
@@ -1374,13 +1374,13 @@
                 if (! mapped)
                 {
                    control_printf(SL_FAILURE, "451 Error: Unable to locate file.");
-                   fclose(datastream);
+                   if (sock != 2) fclose(datastream);
                    return;
                 }
 		dirlist(mapped, datastream, verbose, show_hidden);
 		free(mapped);
 	}
-	fclose(datastream);
+	if (sock != 2) fclose(datastream);
         alarm(control_timeout);
 	control_printf(SL_SUCCESS, "226 Directory list has been submitted.");
 }
--- bftpd/main.c	2022-01-29 01:56:37.000000000 +0800
+++ bftpd-nsfuzz-patch/main.c	2026-04-12 14:58:05.169884261 +0800
@@ -295,7 +295,7 @@
 			 * we have to check if accept() returned an error.
 			 */
 			if (main_sock > 0) {
-				pid = fork();
+				pid = 0;
 				if (!pid) {       /* child */
 					close(0);
 					close(1);
@@ -446,7 +446,9 @@
         alarm(control_timeout);
         
 	/* Read lines from client and execute appropriate commands */
-	while (fgets(str, MAXCMD, stdin)) {
+	while (1) {
+					_NSFUZZ_SYNC();
+					fgets(str, MAXCMD, stdin);
                 int string_length;
                 alarm(control_timeout);
                 string_length = strlen(str) - 2;
